docs/state.rst

```diff
@@ -160,39 +158,39 @@ function.
 
 In order to use this function you need to pass in an integer that represents an application to
 read from. This integer corresponds to the index of an application in the
-:any:`Txn.applications <TxnObject.applications>` array.
-
-This means that in order to read or manipulate an external application's local state, that application
-must be present in the application call transaction's :code:`Txn.applications` array.
+:any:`Txn.applications <TxnObject.applications>` array, or since TEAL version 4, an byte string
```

Comment:
The application argument can never be a byte array. It must be either an integer that's an index into the `Txn.applications` array, or the ID of an application. To avoid confusion, I think it's ok for this page to only mention that the ID can be the real ID of an application. Users can consult the API docs for the functions to learn about the other case.

(Since this is an ambiguous case and app IDs are all large, we actually interpret any integer less than the length of `Txn.applications` as an index, see here for implementation: https://github.com/algorand/go-algorand/blob/3d3feb5183da7b726f5207c18648779d8e9821b3/data/transactions/logic/eval.go#L2993)

Reply:
Good catch. This indicates a mistake both of us overlooked in corresponding `globalPutEx` class, where I let type checking for application id / index to be `anytype`. It should be `uint64`.

I fixed the code section. For document, I think we can just omit the "too-much detailed" information about the argument we passed in.